hpx/lcos/dist_object.hpp

```diff
@@ -0,0 +1,566 @@
+                std::lock_guard<hpx::lcos::local::spinlock> l(lk);
+                servers[source_loc] = id;
+            }
+            b.wait();
+            return servers;
+        }
+
+        HPX_DEFINE_COMPONENT_ACTION(meta_object_server, get_server_list);
+        HPX_DEFINE_COMPONENT_ACTION(meta_object_server, registration);
+
+    private:
+        int num_sent;
+        hpx::lcos::local::barrier b;
+        hpx::lcos::local::spinlock lk;
+        std::vector<hpx::id_type> servers;
+    };
+}}
+typedef hpx::lcos::meta_object_server::get_server_list_action get_list_action;
+HPX_REGISTER_ACTION_DECLARATION(get_list_action, get_server_list_mo_action);
+typedef hpx::lcos::meta_object_server::registration_action
+    register_with_meta_action;
+HPX_REGISTER_ACTION_DECLARATION(register_with_meta_action, register_mo_action);
+
+// Meta_object front end, decides whether it is the root locality, and thus
+// whether to register with the root locality's meta object only or to register
+// itself as the root locality's meta object as well
+namespace hpx { namespace lcos {
+    class meta_object
+      : hpx::components::client_base<meta_object, meta_object_server>
+    {
+    public:
+        typedef hpx::components::client_base<meta_object, meta_object_server>
+            base_type;
+
+        meta_object(std::string basename)
+          : base_type(hpx::local_new<meta_object_server>())
+        {
+            if (hpx::get_locality_id() == 0)
+            {
+                hpx::register_with_basename(
+                    basename, this->get_id(), hpx::get_locality_id());
+            }
+            meta_object_0 = hpx::find_from_basename(basename, 0).get();
+        }
+
+        hpx::future<std::vector<hpx::id_type>> get_server_list()
+        {
+            return hpx::async(get_list_action(), meta_object_0);
+        }
+
+        std::vector<hpx::id_type> registration(hpx::id_type id)
+        {
+            return hpx::async(register_with_meta_action(), meta_object_0,
+                hpx::get_locality_id(), id)
+                .get();
+        }
+
+    private:
+        hpx::id_type meta_object_0;
+    };
+}}
+/// \endcond
+// The front end for the dist_object itself. Essentially wraps actions for
+// the server, and stores information locally about the localities/servers
+// that it needs to know about
+namespace hpx { namespace lcos {
+    /// The dist_object is a single logical object partitioned over a set of
+    /// localities/nodes/machines, where every locality shares the same global
+    /// name locality for the distributed object (i.e. a universal name), but
+    /// owns its local value. In other words, local data of the distributed
+    /// object can be different, but they share access to one another's data
+    /// globally.
+    template <typename T, construction_type C = construction_type::All_to_All>
+    class dist_object
+      : hpx::components::client_base<dist_object<T>,
+            server::dist_object_part<T>>
```

Comment:
With being a client_base, we also get serializability. What does it mean to serialize a `dist_object`?

Reply:
Thanks for the notice. I added code here:
https://github.com/STEllAR-GROUP/hpx/blob/b1c8f32ae0627852cf02f96013a70b443f03a066/hpx/lcos/distributed_object.hpp#L460

Reply:
There is no need to mark this function as `HPX_FORCEINLINE`.

Reply:
@hkaiser removed.